pkg/node/eviction/eviction.go

```diff
@@ -0,0 +1,222 @@
+package eviction
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	policy "k8s.io/api/policy/v1beta1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	listerscorev1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	timeout                   = 60 * time.Second
+	SkipEvictionAnnotationKey = "kubermatic.io/skip-eviction"
+)
+
+type NodeEviction struct {
+	nodeName   string
+	nodeLister listerscorev1.NodeLister
+	client     kubernetes.Interface
+}
+
+// New returns a new NodeEviction
+func New(nodeName string, nodeLister listerscorev1.NodeLister, client kubernetes.Interface) *NodeEviction {
+	return &NodeEviction{
+		nodeName:   nodeName,
+		nodeLister: nodeLister,
+		client:     client,
+	}
+}
+
+// Run excutes the eviction
+func (ne *NodeEviction) Run() error {
+	listerNode, err := ne.nodeLister.Get(ne.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node from lister: %v", err)
+	}
+	node := listerNode.DeepCopy()
+	if node.Annotations != nil {
+		if _, exists := node.Annotations[SkipEvictionAnnotationKey]; exists {
+			glog.V(4).Infof("Skipping eviction for node %s as it has a %s annotation", ne.nodeName, SkipEvictionAnnotationKey)
+			return nil
+		}
+	}
+	glog.V(4).Infof("Starting to evict node %s", ne.nodeLister)
+
+	if err := ne.cordonNode(node); err != nil {
+		return fmt.Errorf("failed to cordon node %s: %v", ne.nodeName, err)
+	}
+	glog.V(4).Infof("Successfully cordoned node %s", ne.nodeName)
+
+	podsToEvict, err := ne.getFilteredPods()
+	if err != nil {
+		return fmt.Errorf("failed to get Pods to evict for node %s: %v", ne.nodeName, err)
+	}
+	glog.V(4).Infof("Found %v pods to evict for node %s", len(podsToEvict), ne.nodeName)
+
+	if errs := ne.evictPods(podsToEvict); len(errs) > 0 {
+		return fmt.Errorf("failed to evict pods, errors encountered: %v", errs)
+	}
+	glog.V(4).Infof("Successfully evicted all pods for node %s!", ne.nodeName)
+
+	glog.V(4).Infof("Waiting for deletion of all pods for node %s", ne.nodeName)
+	if err := ne.waitForDeletion(podsToEvict); err != nil {
+		return fmt.Errorf("failed waiting for pods of node %s to be deleted: %v", ne.nodeName, err)
+	}
+	glog.V(4).Infof("All pods of node %s were successfully deleted", ne.nodeName)
+
+	return nil
+}
+
+func (ne *NodeEviction) cordonNode(node *corev1.Node) error {
+	_, err := ne.updateNode(func(n *corev1.Node) {
+		n.Spec.Unschedulable = true
+	})
+	if err != nil {
+		return err
+	}
+
+	// Be paranoid and wait until the change got propagated to the lister
```

Comment:
Why exactly here?

Reply:
Because its after the node got set to `Unschedulable` and before we call `getFilteredPods` the purpose is to be really sure that the scheduler does not schedule any pods in between setting this and after `getFilteredPods` because it doesn't have the change yet

Reply:
Yes I know that this might not be 100% accurate, however the chances of a node getting scheduled in exactly this timeframe is pretty low. I would prefer not to add a 5s sleep there, but will add a comment